### PR TITLE
Enable stronger cryptnono protections for public hubs + allow sharing shared directories between hubs

### PR DESCRIPTION
**Pull request by yuvipanda**
- Bring in last 6 months of work on cryptnono for mybinder.org into this repo, and document it heavily. Read the documentation for more info!
- Enable stronger cryptnono protections for HHMI spyglass ephemeral
- Move HHMI spyglass ephemeral hub to using tmpauthenticator open to the world instead of CILogon
- Document that ephemeral hubs should have this extra protection be enabled
- Turns out a core part of the HHMI demonstration involves having shared directories with ~40G of
   data in them. After brainstorming a few solutions, we've currently settled on sharing the shared  
   directory between the two hubs in this cluster (HHMI prod as well as spyglass ephemeral). There is
   documentation added about doing this as well.
- I was asked to bump an image tag for this by @jmunroe and have done so.

Ref https://github.com/2i2c-org/infrastructure/issues/3643

### CODE REVIEW

File: config/clusters/hhmi/spyglass.values.yaml [hunk 1]
```diff
@@ -41,14 +50,20 @@ jupyterhub:
...
+        - name: shared-dir-pv
+          mountPath: /home/jovyan/shared-readonly
+          subPath: _shared
+          readOnly: true
     defaultUrl: /lab
     image:
-      name: quay.io/lorenlab/hhmi-spyglass-image
-      tag: "c307f9418a60"
+      name: quay.io/2i2c/hhmi-spyglass-image
+      tag: "8be848d09076"
```
Comment (consideRatio) [inline config/clusters/hhmi/spyglass.values.yaml:66]: Unrelated change - I think it may be reverting something new for something old - right?

File: config/clusters/hhmi/spyglass.values.yaml [hunk 2]
```diff
@@ -72,12 +87,9 @@ jupyterhub:
...
-        authenticator_class: cilogon
-      CILogonOAuthenticator:
-        oauth_callback_url: https://spyglass.hhmi.2i2c.cloud/hub/oauth_callback
-        allowed_idps:
-          http://github.com/login/oauth/authorize:
-            default: true
-            username_derivation:
-              username_claim: "preferred_username"
-            allow_all: true
+        authenticator_class: tmpauthenticator.TmpAuthenticator
```
Comment (consideRatio) [inline config/clusters/hhmi/spyglass.values.yaml:90]:
```suggestion
        authenticator_class: tmp
```

Added in tmpauthenticator 1.0.0 via https://github.com/jupyterhub/tmpauthenticator/commit/f2f6deed9962223a7386b53d2ae5c98844195f06

File: config/clusters/hhmi/spyglass.values.yaml (hunk 1 above)
Comment (yuvipanda) [reply config/clusters/hhmi/spyglass.values.yaml:66]: @consideRatio yes, unrelated but necessary - I put this in the description:

> I was asked to bump an image tag for this by @jmunroe and have done so.

I couldn't keep it separate because it was needed for him to test his changes.

File: deployer/commands/generate/cryptnono_config/__init__.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+HERE = Path(__file__).parent
+REPO_ROOT_PATH = HERE.parent.parent.parent.parent
+
+
+@generate_app.command()
+def cryptnono_secret_config():
+    """
+    Update the secret blocklist for cryptnono
+    """
+    unencrypted_path = HERE / "unencrypted_secret_blocklist.py"
```
Comment (consideRatio) [inline deployer/commands/generate/cryptnono_config/__init__.py:16]: This should be following the pattern on how to name encrypted / decrypted conten to make our .gitignore file protect us from adding it unencrypted.

```gitignore
# This section handles sensitive content that is or isn't encrypted in our repo.
#
# Ignore files (but not directories) with secret in their name
*secret*
!*secret*/
# Don't ignore enc- prefixed files
!enc-*
# Don't ignore Helm chart templates with secret in their name as
# they are not supposed to contain sensitive content but can often
# have secret in their name.
!**/templates/**/*secret*.yaml
```

Let's make `enc-<name>.secret.<file extension>` to `<name>.secret.<file extension>`.

```suggestion
    unencrypted_path = HERE / "blocklist-generator.secret.py"
```

File: docs/howto/features/cryptnono.md [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+deployer generate cryptnono-secret-config
+```
+
+This will update the file `helm-charts/support/enc-cryptnono.secret.values.yaml` and re-encrypt it.
+
+## Working on the banned strings generator
+
+The banned strings generator is a fairly simple python script, present in `deployer/commands/generate/cryptnono_config/enc-blocklist-generator.secret.py`.
+It's unencrypted and loaded by code in `deployer/commands/generate/cryptnono_config/__init__.py`. There is inline
+documentation in `encrypted_secret_blocklist.py`, but how does one hack on it?
```
Comment (consideRatio) [inline docs/howto/features/cryptnono.md:134]:
```suggestion
documentation in `enc-blocklist-generator.secret.py`, but how does one maintain it?
```

File: docs/howto/features/cryptnono.md [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+## Detectors
+
+Cryptnono currently has two primary detectors:
+
+1. A detector for the [monero](https://www.getmonero.org/) cryptocurrency, that is based on
+   official guidance [from the monero project](https://blog.px.dev/detect-monero-miners/) on
+   how to detect it. This is fairly safe and has a very low false positive rate, and requires
+   no configuration. So by default, **this detector is enabled on all hubs**.
+
+2. A detector (`execWhacker`) based on heuristics, where the full commandline used to execute a process (
```
Comment (consideRatio) [inline docs/howto/features/cryptnono.md:23]: This is capitalized as `execWhacker` but the chart accept `execwhacker` rather than `execWhacker`.

I think it would be good to consistently refer to one capitalization found in the chart config, or something clearly distinguished as written for human eyes, like `the exec whacker`.

File: deployer/commands/generate/cryptnono_config/__init__.py [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+from deployer.cli_app import generate_app
+
+HERE = Path(__file__).parent
+REPO_ROOT_PATH = HERE.parent.parent.parent.parent
+
+
+@generate_app.command()
+def cryptnono_secret_config():
+    """
+    Update the secret blocklist for cryptnono
```
Comment (consideRatio) [inline deployer/commands/generate/cryptnono_config/__init__.py:14]: I wanted to make this code and command not require pre-requisite knowledge on what this what about, but allow the help string (from the docstring) to convey that.

```suggestion
    Updates secret config for the cryptnono chart bundled with the support chart, specifically banned command strings and patterns as required by cryptnono's execwhacker detector if its enabled.
    
    The banned strings are generated using a secret script, and the deployer reads the secret config when deploying the support chart.
```

File: config/clusters/hhmi/spyglass.values.yaml [hunk 7]
```diff
@@ -72,12 +87,9 @@ jupyterhub:
...
+      TmpAuthenticator:
+        # This allows users to go to the hub URL directly again to
+        # get a new server, instead of being plopped back into their
+        # older, existing user with a 'start server' button.
+        force_new_server: true
```
Comment (consideRatio) [inline config/clusters/hhmi/spyglass.values.yaml:95]: This is no longer needed, see https://github.com/jupyterhub/tmpauthenticator/blob/main/CHANGELOG.md#breaking-changes.

```suggestion
```

File: docs/howto/features/cryptnono.md [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+   look for in the commandline of a process, and is [constantly being tweaked](https://github.com/jupyterhub/mybinder.org-deploy/security/advisories/GHSA-j42g-x8qw-jjfh).
+   Since making the list of banned strings public would make ban evasion easy, the list of
+   strings (and the method used to generate them) is encrypted. You can read more details
+   about the specific method used by looking in the encrypted code file (`deployer/commands/generate/cryptnono_config/encrypted_secret_blocklist.py`)
+   in this repository.
+
+   Since this detector may have a non-0 false positive rate, it is currently *not* enabled by
+   default. However, eventually, once the config matures enough (and we have tested it enough),
+   this would also be enabled by default everywhere. In the meantime, we only enable it for
+   *clusters* with any hub that allows unfettered external access.
```
Comment (consideRatio) [inline docs/howto/features/cryptnono.md:35]: I opened https://github.com/cryptnono/cryptnono/issues/31 about this. I think a criteria for 2i2c to enable execwhacker by default should be that there are metrics or logs to help us rule out or conclude cryptnono's execwhacker killed a process.

File: docs/howto/features/cryptnono.md [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+   ```
+
+   The `-o wide` will add an additional column, `NODE`, showing which node the pods are running in. Find the
+   node of the user pod you care about.
+
+2. Find the appropriate `cryptnono` pod for this node.
+
+   ```bash
+   kubectl -n support get pod \
+    --field-selector spec.nodeName=<name-of-node>\
```
Comment (consideRatio) [inline docs/howto/features/cryptnono.md:90]:
```suggestion
    --field-selector spec.nodeName=<name-of-node> \
```

File: config/clusters/hhmi/spyglass.values.yaml (hunk 1 above)
Comment (consideRatio) [inline config/clusters/hhmi/spyglass.values.yaml:66]:
```suggestion
      name: quay.io/lorenlab/hhmi-spyglass-image	      name: quay.io/2i2c/hhmi-spyglass-image
      tag: "c307f9418a60"
```

Comment (consideRatio) [inline config/clusters/hhmi/spyglass.values.yaml:66]: I think maybe the entire set of changes to spyglass.values.yaml are unrelated?

File: config/clusters/hhmi/spyglass.values.yaml [hunk 10]
```diff
@@ -1,7 +1,16 @@
...
+  dirsizeReporter:
     enabled: false
+  pv:
+    enabled: true
+    mountOptions:
+      - soft
+      - noatime
+    serverIP: 10.55.112.74
+    baseShareName: /homes/
+    shareNameOverride: prod
```
Comment (consideRatio) [inline config/clusters/hhmi/spyglass.values.yaml:12]:
```suggestion
    # the spyglass hub and the prod hub is configured to use the same storage!
    shareNameOverride: prod
```

File: docs/howto/features/ephemeral.md [hunk 11]
```diff
@@ -109,6 +69,77 @@ jupyterhub:
...
+          - name: shared-dir-pvc
+            persistentVolumeClaim:
+              # The name of the PVC setup by nfs.yaml for the ephemeral hub to use
+              claimName: home-nfs
+        extraVolumeMounts:
+          - name: shared-dir-pvc
+            mountPath: /home/jovyan/shared
+            subPath: _shared
+            readOnly: true
+   ```
```
Comment (consideRatio) [inline docs/howto/features/ephemeral.md:136]: Indentation fix, adding a space for all lines after 122.

````suggestion
   ```yaml
   jupyterhub:
     singleuser:
       storage:
         # We still don't want to have per-user storage
         type: none
         extraVolumes:
           - name: shared-dir-pvc
             persistentVolumeClaim:
               # The name of the PVC setup by nfs.yaml for the ephemeral hub to use
               claimName: home-nfs
         extraVolumeMounts:
           - name: shared-dir-pvc
             mountPath: /home/jovyan/shared
             subPath: _shared
             readOnly: true
   ```
````

File: docs/howto/features/ephemeral.md [hunk 12]
```diff
@@ -170,8 +201,67 @@ jupyterhub:
...
+
+```yaml
+ingressBasicAuth:
+  enabled: true
+
+jupyterhub:
+  ingress:
+    annotations:
+      # We protect our entire hub from cryptobros by putting it all
+      # behind a single shared basicauth
```
Comment (consideRatio) [inline docs/howto/features/ephemeral.md:243]:
```suggestion
      # We protect our entire hub from cryptominers by putting it all
      # behind a single shared basicauth
```

Its subjective of course, but I appreciate a more neutral language - wdyt?

File: helm-charts/support/values.yaml [hunk 13]
```diff
@@ -422,6 +422,12 @@ cryptnono:
     # Disable the execwhacker detector for now, as it matures by being deployed on mybinder.org
     execwhacker:
       enabled: false
+      configs:
+        unencrypted-test-01:
+          bannedCommandStrings:
+            # Provide an unencrypted, randomly generated test string. All processes that have this in their commandline will be killed.
+            # This is helpful for testing by engineers
```
Comment (consideRatio) [inline helm-charts/support/values.yaml:429]:
```suggestion
            # This is helpful for testing execwhacker is enabled and functional
```

**Review by consideRatio — APPROVED** [review]
I made some comments, but this looks overall great to me so I'll mark it as approved! Great seeing that there are already metrics and nice logging in place in cryptnono's execwhacker!

Amazing work on this @yuvipanda!!

File: deployer/commands/generate/cryptnono_config/__init__.py [hunk 14]
```diff
@@ -0,0 +1,34 @@
+import shutil
+import subprocess
+from pathlib import Path
+
+from deployer.cli_app import generate_app
+
+HERE = Path(__file__).parent
+REPO_ROOT_PATH = HERE.parent.parent.parent.parent
```
Comment (GeorgianaElena) [inline deployer/commands/generate/cryptnono_config/__init__.py:8]: I think we should import the one in utils so that if we ever want to restructure the deployer again or relocate this file, we don't need to update it everywhere, just in one place.

```suggestion
from deployer.cli_app import generate_app
from deployer.utils.file_acquisition import REPO_ROOT_PATH

HERE = Path(__file__).parent
```

File: docs/howto/features/cryptnono.md [hunk 15]
```diff
@@ -0,0 +1,176 @@
...
+2. The secret config itself.
+
+If a community wishes to leave and needs the config, we can make sure they know what the config is and
+how to keep it updated - very similar to how we would handle passing on CILogon credentials or similar to
+them. Since none of the code for cryptnono itself is secret, this does not conflict with right to replicate.
+
+## Future work
+
+`cryptnono` also exposes prometheus metrics about processes it has killed. We currently do not collect these,
+but we should enable collection so we can
```
Comment (GeorgianaElena) [inline docs/howto/features/cryptnono.md:176]: I think it misses a few closing words :D

File: helm-charts/basehub/values.schema.yaml [hunk 16]
```diff
@@ -213,6 +213,11 @@ properties:
             type: string
           baseShareName:
             type: string
+          shareNameOverride:
+            type: string
+            descirption: |
```
Comment (GeorgianaElena) [inline helm-charts/basehub/values.schema.yaml:218]: Small typo

```suggestion
            description: |
```

**Review by GeorgianaElena — APPROVED** [review]
Amazing work @yuvipanda 🚀 Thank you!

File: docs/howto/features/cryptnono.md [hunk 17]
```diff
@@ -0,0 +1,176 @@
...
+   no configuration. So by default, **this detector is enabled on all hubs**.
+
+2. A detector (`execWhacker`) based on heuristics, where the full commandline used to execute a process (
+   regardless of how it is started) is used to detect if a process is likely crypto mining,
+   and if so, immediately kill it. This relies on a tweakable config of banned strings to
+   look for in the commandline of a process, and is [constantly being tweaked](https://github.com/jupyterhub/mybinder.org-deploy/security/advisories/GHSA-j42g-x8qw-jjfh).
+   Since making the list of banned strings public would make ban evasion easy, the list of
+   strings (and the method used to generate them) is encrypted. You can read more details
+   about the specific method used by looking in the encrypted code file (`deployer/commands/generate/cryptnono_config/encrypted_secret_blocklist.py`)
+   in this repository.
```
Comment (sgibson91) [inline docs/howto/features/cryptnono.md:30]:
```suggestion
   in the `infrastructure` repository.
```

I don't think it's clear when you're on https://infrastructure.2i2c.org which repository 'this' one is.

**Comment by sgibson91** [comment]
@yuvipanda I opened https://github.com/yuvipanda/pilot-hubs/pull/31 with a suggested splitting of the docs to follow diataxis a bit more

File: deployer/commands/generate/cryptnono_config/__init__.py (hunk 3 above)
Comment (yuvipanda) [reply deployer/commands/generate/cryptnono_config/__init__.py:16]: The paths already take into account our gitignore. The encrypted file to be checked in is called `enc-something.secret.py` so it can be checked in. Our gitignore ignores anything with secret but without enc, so this temporary unencrypted file *will* be ignored. We can't use dashes in there because then the python import statement would not work, and the same is true of additional dots in the filename. As an additional precaution, the `finally` block deletes this file as well.

File: docs/howto/features/cryptnono.md (hunk 5 above)
Comment (yuvipanda) [reply docs/howto/features/cryptnono.md:23]: Yes, I'll do that. I think when writing cryptnono I initially had called it `execWhacker` before moving to all lowercase - I'll think about what's the most appropriate casing, and unify that (including possibly changing it in cryptnono itself).

File: docs/howto/features/ephemeral.md (hunk 12 above)
Comment (yuvipanda) [reply docs/howto/features/ephemeral.md:243]: Ah, this is a rabbithole that I'd prefer to not get into in this PR. Is it alright if we just let this one be for now?

File: deployer/commands/generate/cryptnono_config/__init__.py (hunk 6 above)
Comment (yuvipanda) [reply deployer/commands/generate/cryptnono_config/__init__.py:14]: This feels a little too repetitive for my personal style. Instead, once the PR is merged, I'll link to the docs from the code comments here - I put a lot of effort into those, and I think that's a better place for this kind of information.

File: docs/howto/features/ephemeral.md (hunk 11 above)
Comment (yuvipanda) [reply docs/howto/features/ephemeral.md:136]: I've no idea how you caught that but *great* catch.

File: docs/howto/features/cryptnono.md (hunk 8 above)
Comment (yuvipanda) [reply docs/howto/features/cryptnono.md:35]: @consideRatio indeed! I've now enabled container information reporting as well, so we will know what pod / namespace the kill took place in. We're also now collecting the metrics that cryptnono exposes, even though there is no grafana dashboard for that yet.

Comment (yuvipanda) [reply docs/howto/features/cryptnono.md:35]: I've updated the docs to match this, including a sample log entry.

File: docs/howto/features/cryptnono.md (hunk 5 above)
Comment (yuvipanda) [reply docs/howto/features/cryptnono.md]: I've just moved everything to say `execwhacker` for now, I'll update it here if I change it in cryptnono

File: config/clusters/hhmi/spyglass.values.yaml (hunk 10 above)
Comment (yuvipanda) [reply config/clusters/hhmi/spyglass.values.yaml:12]: I've added an appropriate comment here!

File: docs/howto/features/cryptnono.md (hunk 15 above)
Comment (yuvipanda) [reply docs/howto/features/cryptnono.md]: Ah yep! I've actually enabled collection of these now, and updated this section to say that we just don't have a grafana board yet.

**Comment by yuvipanda** [comment]
Thanks for the review everyone! I'm going to merge this now so @jmunroe can make image tag bumps without relying on me. I have two follow-ups here:

- [ ] The `tmpauthenticator` related changes that @consideRatio pointed out. I'll make a separate PR, that also changes the docs appropriately
- [ ] The wonderful split that @sgibson91 had done in https://github.com/yuvipanda/pilot-hubs/pull/31. Some conflicts need to be resolved, and I'll try to pick that up too.

Thanks!

File: config/clusters/hhmi/spyglass.values.yaml (hunk 2 above)
Comment (yuvipanda) [reply config/clusters/hhmi/spyglass.values.yaml:90]: Done in https://github.com/2i2c-org/infrastructure/pull/3669

File: config/clusters/hhmi/spyglass.values.yaml (hunk 7 above)
Comment (yuvipanda) [reply config/clusters/hhmi/spyglass.values.yaml:95]: Done in https://github.com/2i2c-org/infrastructure/pull/3669
